pydantic/generics.py

```diff
@@ -0,0 +1,82 @@
+from typing import Any, ClassVar, Dict, Generic, Tuple, Type, TypeVar, Union, get_type_hints
+
+from pydantic import BaseModel, create_model
+from pydantic.class_validators import gather_validators
+
+_generic_types_cache: Dict[Tuple[Type[Any], Union[Any, Tuple[Any, ...]]], Type[BaseModel]] = {}
+GenericModelT = TypeVar('GenericModelT', bound='GenericModel')
+
+
+class GenericModel(BaseModel):
+    __slots__ = ()
+    __concrete: ClassVar[bool] = False
+
+    def __new__(cls, *args: Any, **kwargs: Any) -> GenericModelT:
+        if cls.__concrete:
+            return super().__new__(cls)
+        if Generic in cls.__bases__:
+            raise TypeError(f'Type {cls.__name__} cannot be instantiated without providing generic parameters')
+        else:
+            raise TypeError(f'Type {cls.__name__} cannot be instantiated; it can be used only as a base class')
+
+    def __class_getitem__(  # type: ignore
+        cls: Type[GenericModelT], params: Union[Type[Any], Tuple[Type[Any], ...]]
+    ) -> Type[BaseModel]:
+        cached = _generic_types_cache.get((cls, params))
+        if cached is not None:
+            return cached
+
+        if not isinstance(params, tuple):
+            params = (params,)
+        if any(isinstance(param, TypeVar) for param in params):  # type: ignore
+            raise TypeError(f'Type parameters should be placed on typing.Generic, not GenericModel')
+
+        check_parameters_count(cls, params)
+        typevars_map: Dict[Any, Any] = dict(zip(cls.__parameters__, params))  # type: ignore
+        type_hints = get_type_hints(cls).items()
+        instance_type_hints = {k: v for k, v in type_hints if getattr(v, "__origin__", None) is not ClassVar}
+        concrete_type_hints: Dict[str, Type[Any]] = {
+            k: resolve_type_hint(v, typevars_map) for k, v in instance_type_hints.items()
+        }
+
+        model_name = concrete_name(cls, params)
+        validators = gather_validators(cls)
+        fields: Dict[str, Tuple[Type[Any], Any]] = {
+            k: (v, getattr(cls, k, ...)) for k, v in concrete_type_hints.items()
+        }
+        created_model = create_model(
+            model_name=model_name,
+            __module__=cls.__module__,
+            __base__=cls,
+            __config__=None,
+            __validators__=validators,
+            **fields,
+        )
+        created_model.Config = cls.Config
```

Comment:
I got everything working now so that methods defined on the generic are inherited properly, etc. (See tests for more clarity). The `create_model` function doesn't let you specify both a config and a base class though, so I resorted to manually setting it afterward.

Since `created_model` is a concrete instantiation, I think it makes sense to have all subclasses actually share the *same* config instance, but let me know if you think it would be better to create a copy.